####################################################
# Snakemake rules
####################################################

def create_multiqc_bam_individual_input(wildcards):

    species = wildcards.species
    reference = wildcards.reference
    individual = wildcards.individual

    file_list = []

    #for individual in individuals:
    # get samples for individual
    samples_of_individual = get_samples_for_species_individual(species, individual)

    for sample in samples_of_individual:

        # get raw read file paths
        raw_reads = get_raw_reads_for_sample(species, sample)

        if config.get("pipeline", {}).get("reads_processing", {}).get("adapter_removal", {}).get("execute", True) == True:
            # add fastp json reports
            # fastp trimming reports
            if len(raw_reads) == 2:
                file_list.append(f"{species}/results/reads/reads_trimmed/fastp_report/{sample}_trimmed.pe.json")
            else:
                file_list.append(f"{species}/results/reads/reads_trimmed/fastp_report/{sample}_trimmed.se.json")

        if config.get("pipeline", {}).get("reads_processing", {}).get("quality_filtering", {}).get("execute", True) == True:
            # fastp quality filtering reports
            file_list.append(f"{species}/results/reads/reads_quality_filtered/fastp_report/{sample}_quality_filtered.json")


        # contamination analysis outputs
        if config.get("pipeline", {}).get("reference_processing", {}).get("contamination_analysis", {}).get("execute", True) == True:

            if config.get("pipeline", {}).get("reference_processing", {}).get("contamination_analysis", {}).get("tools", {}).get("ecmsd", {}).get("execute", True) == True:
                file_list.append(f"{species}/results/summary/{individual}/multiqc_custom_content/{sample}/{sample}_Mito_summary_genus_ReadLengths.png")

            if config.get("pipeline", {}).get("reference_processing", {}).get("contamination_analysis", {}).get("tools", {}).get("centrifuge", {}).get("execute", True) == True:
                #file_list.append(f"{species}/results/contamination_analysis/centrifuge/{individual}/{sample}/{sample}_centrifuge_report.tsv")
                file_list.append(f"{species}/results/contamination_analysis/centrifuge/{individual}/{sample}/{sample}_top10_unique_taxa.tsv")
                file_list.append(f"{species}/results/contamination_analysis/centrifuge/{individual}/{sample}/{sample}_top10_total_taxa.tsv")

        
    # merged reads fastqc
    if config.get("pipeline", {}).get("reads_processing", {}).get("quality_checking_merged", {}).get("execute", True) == True:
        file_list.append(f"{species}/results/reads/reads_merged/fastqc/{individual}_merged_fastqc.zip")

    if config.get("pipeline", {}).get("reference_processing", {}).get("coverage_analysis", {}).get("execute", True) == True:
        #file_list.append(f"{species}/results/{reference}/analytics/{individual}/preseq/{individual}_{reference}.lc_extrap")
        file_list.append(f"{species}/results/{reference}/analytics/{individual}/preseq/{individual}_{reference}.c_curve.txt")
        file_list.append(directory(f"{species}/results/{reference}/analytics/{individual}/qualimap"))
        file_list.append(f"{species}/results/{reference}/analytics/{individual}/samtools_stats/{individual}_{reference}_final.bam.stats")
        file_list.append(f"{species}/results/summary/{individual}/multiqc_custom_content/{individual}_{reference}_reads_processing_summary.tsv")
        file_list.append(f"{species}/results/summary/{individual}/multiqc_custom_content/{individual}_{reference}_coverage_analysis.tsv")
        file_list.append(f"{species}/results/summary/{individual}/multiqc_custom_content/{individual}_{reference}_depth_coverage_avg.csv")
        file_list.append(f"{species}/results/summary/{individual}/multiqc_custom_content/{individual}_{reference}_coverage_summary.tsv")
    
    if config.get("pipeline", {}).get("reference_processing", {}).get("damage_rescaling", {}).get("execute", True) == True:
        file_list.append(directory(f"{species}/results/{reference}/analytics/{individual}/mapdamage/"))
    
    if config.get("pipeline", {}).get("reference_processing", {}).get("deduplication", {}).get("execute", True) == True:
        file_list.append(f"{species}/results/{reference}/analytics/{individual}/dedup/{individual}_{reference}_final.dedup.json")

    return file_list

####################################################
# Snakemake rules
####################################################
rule create_multiqc_bam_individual:
    input:
        create_multiqc_bam_individual_input,
        config="{species}/results/summary/{individual}/{individual}_{reference}_multiqc_config.yaml"
    output:
        "{species}/results/{reference}/analytics/{individual}_{reference}_multiqc.html",
        directory("{species}/results/{reference}/analytics/{individual}/multiqc_data"),
    params:
        extra="--verbose",  # Optional: extra parameters for multiqc.
    log:
        "{species}/results/{reference}/analytics/{individual}/multiqc.log",
    wrapper:
        "v7.9.0/bio/multiqc"

# rule create_multiqc_folder:
#     input:
#         insurance = "{species}/results/{reference}/analytics/request_multiqc_bam.ready"
#     output:
#         temp(directory("{species}/results/{reference}/multiqc/"))
#     shell:
#         "mkdir -p {output}"

# rule validate_multiqc_preprocessing:
#     input:
#         #picard_metrics = "{species}/results/{reference}/analytics/{individual}/picard_duplicates/{individual}_{reference}_metrics.txt",
#         preseq_lc_extrap = "{species}/results/{reference}/analytics/{individual}/preseq/{individual}_{reference}.lc_extrap",
#         samtools_stats = "{species}/results/{reference}/analytics/{individual}/samtools_stats/{individual}_{reference}_final.bam.stats",
#         qualimap = "{species}/results/{reference}/analytics/{individual}/qualimap",
#     output:
#         preprocessing = temp("{species}/results/{reference}/analytics/{individual}/multiqc_preprocessing.ready")
#     shell:
#         "touch {output}"
    
# rule request_multiqc_bam:
#     input:
#         preprocessing = lambda wildcards: expand(
#             "{species}/results/{reference}/analytics/{individual}/multiqc_preprocessing.ready",
#             species     = wildcards.species,
#             reference   = wildcards.reference,
#             individual  = get_individuals_for_species(wildcards.species),
#         )
#     output:
#         ok_file = temp("{species}/results/{reference}/analytics/request_multiqc_bam.ready")
#     shell:
#         """
#         echo 'Requesting preprocessing for multiqc report:'
        
#         for f in {input.preprocessing}; do
#             echo $f
#         done

#         touch {output.ok_file}
#         """

rule create_multiqc_bam_individual_config:
    output:
        "{species}/results/summary/{individual}/{individual}_{reference}_multiqc_config.yaml"
    script:
        "../../../scripts/processing_summary/create_multiqc_species_individual_script_create_multiqc_species_individual_config.py"
        